src/Raven.Server/Https/ExternalCertificateValidator.cs

```diff
@@ -0,0 +1,288 @@
+            }
+
+            try
+            {
+                readStdOut.Wait(timeout);
+                readErrors.Wait(timeout);
+            }
+            catch (Exception e)
+            {
+                throw new InvalidOperationException($"Unable to read redirected stderr and stdout when executing '{_server.Configuration.Security.CertificateValidationExec} {args}'", e);
+            }
+
+            string output = GetStdOut();
+            string errors = GetStdError();
+
+            // Can have exit code 0 (success) but still get errors. We log the errors anyway.
+            if (log.IsOperationsEnabled)
+                log.Operations($"Executing '{_server.Configuration.Security.CertificateValidationExec} {args}' took {sw.ElapsedMilliseconds:#,#;;0} ms. Exit code: {process.ExitCode}{Environment.NewLine}Output: {output}{Environment.NewLine}Errors: {errors}{Environment.NewLine}");
+
+            if (process.ExitCode != 0)
+            {
+                throw new InvalidOperationException(
+                    $"Command or executable '{_server.Configuration.Security.CertificateValidationExec} {args}' failed. Exit code: {process.ExitCode}{Environment.NewLine}Output: {output}{Environment.NewLine}Errors: {errors}{Environment.NewLine}");
+            }
+
+            if (bool.TryParse(output, out bool result) == false)
+            {
+                throw new InvalidOperationException(
+                    $"Cannot parse to boolean the result of Command or executable '{_server.Configuration.Security.CertificateValidationExec} {args}'. Exit code: {process.ExitCode}{Environment.NewLine}Output: {output}{Environment.NewLine}Errors: {errors}{Environment.NewLine}");
+            }
+
+            return new CachedValue { Valid = result, Until = result ? DateTime.UtcNow.AddMinutes(15) : DateTime.UtcNow.AddSeconds(30) };
+        }
+
+        public bool ExternalCertificateValidationCallback(object sender, X509Certificate certificate, X509Chain chain, SslPolicyErrors sslPolicyErrors, Logger log)
+        {
+            var senderHostname = RequestExecutor.ConvertSenderObjectToHostname(sender);
+            var lastWriteTimeHash = GetScriptLastWriteHash();
+            var cacheKey = new Key(senderHostname, certificate.GetCertHashString(), sslPolicyErrors, lastWriteTimeHash);
+
+            Task<CachedValue> task;
+            if (_externalCertificateValidationCallbackCache.TryGetValue(cacheKey, out var existingTask) == false)
+            {
+                task = new Task<CachedValue>(() => CheckExternalCertificateValidation(senderHostname, certificate, chain, sslPolicyErrors, log));
+                existingTask = _externalCertificateValidationCallbackCache.GetOrAdd(cacheKey, task);
+                if (existingTask == task)
+                {
+                    task.Start();
+
+                    if (_externalCertificateValidationCallbackCache.Count > 50)
```

Comment:
Shouldn't we clear that not only when we do not have given key in cache?

Reply:
No. When we don't have a key in the cache is when we add a new item to the cache which causes it to grow.

When a key exists we just use the cached value and return quickly